src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -698,4 +576,260 @@ public SNIError GetLastError()
+
+            if (details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            if (!details.InferConnectionDetails() && details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            return details;
+        }
+
+        private bool InferLocalServerName()
+        {
+            // If Server name is empty or localhost, then use "localhost"
+            if (string.IsNullOrEmpty(ServerName) || IsLocalHost(ServerName))
+            {
+                ServerName = ConnectionProtocol == DataSource.Protocol.Admin ?
+                    Environment.MachineName : DefaultHostName;
+            }
+            return true;
+        }
+
+        private bool InferConnectionDetails()
+        {
+            string[] tokensByCommaAndSlash = _dataSourceAfterTrimmingProtocol.Split(BackSlashSeparator, ',');
+            ServerName = tokensByCommaAndSlash[0];
+
+            int commaIndex = _dataSourceAfterTrimmingProtocol.IndexOf(',');
+
+            int backSlashIndex = _dataSourceAfterTrimmingProtocol.IndexOf(BackSlashSeparator);
+
+            // Check the parameters. The parameters are Comma separated in the Data Source. The parameter we really care about is the port
+            // If Comma exists, the try to get the port number
+            if (commaIndex > -1)
+            {
+                bool commaAfterSlash = commaIndex > backSlashIndex ? true : false;
+
+                string parameter = backSlashIndex > -1
+                        ? commaIndex > backSlashIndex ? tokensByCommaAndSlash[2].Trim() : tokensByCommaAndSlash[1].Trim()
+                        : tokensByCommaAndSlash[1].Trim();
+
+                // Bad Data Source like "server, "
+                if (string.IsNullOrEmpty(parameter))
+                {
+                    return ReportSNIError(SNIProviders.INVALID_PROV);
+                }
+
+                // For Tcp and Only Tcp are parameters allowed.
+                if (ConnectionProtocol == DataSource.Protocol.None)
+                {
+                    ConnectionProtocol = DataSource.Protocol.TCP;
+                }
+                else if (ConnectionProtocol != DataSource.Protocol.TCP)
+                {
+                    // Parameter has been specified for non-TCP protocol. This is not allowed.
+                    return ReportSNIError(SNIProviders.INVALID_PROV);
+                }
+
+                int port;
+                if (!int.TryParse(parameter, out port))
+                {
+                    return ReportSNIError(SNIProviders.TCP_PROV);
+                }
+
+                Port = port;
+            }
+
+            // Instance Name Handling. Only if we found a '\' and we did not find a port in the Data Source
+            if (backSlashIndex > -1 && Port == -1)
+            {
+                // This means that there will not be any part separated by comma. 
+                InstanceName = tokensByCommaAndSlash[1].Trim();
```

Comment:
Also the networking layer provides more details information. 
E.g. Here is a failure to connect to Sql Server on localhost,66777 with these changes.


       Actual:   System.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurre
 d while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the inst
 ance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 3
 5 - An internal exception was caught) ---> System.AggregateException: One or more errors occurred. (Specified argumen
 t was out of the range of valid values.
       Parameter name: port) ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid
  values.
       Parameter name: port
          at System.Net.Sockets.Socket.BeginConnect(IPAddress address, Int32 port, AsyncCallback requestCallback, Obje
 ct state) in e:\work\gitclone\corefx\src\System.Net.Sockets\src\System\Net\Sockets\Socket.cs:line 2248
          at System.Net.Sockets.Socket.ConnectAsync(IPAddress address, Int32 port) in e:\work\gitclone\corefx\src\Syst
 em.Net.Sockets\src\System\Net\Sockets\Socket.Tasks.cs:line 70
          at System.Net.Sockets.SocketTaskExtensions.ConnectAsync(Socket socket, IPAddress address, Int32 port) in e:\
 work\gitclone\corefx\src\System.Net.Sockets\src\System\Net\Sockets\SocketTaskExtensions.cs:line 20
          at System.Data.SqlClient.SNI.SNITCPHandle.ParallelConnectAsync(IPAddress[] serverAddresses, Int32 port) in E
 :\work\gitclone\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNITcpHandle.cs:line 237
       --- End of stack trace from previous location where exception was thrown ---
          at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
          at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
          at System.Data.SqlClient.SNI.SNITCPHandle.<ParallelConnectHelper>d__24.MoveNext() in E:\work\gitclone\corefx
 \src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNITcpHandle.cs:line 265
       --- End of stack trace from previous location where exception was thrown ---
          at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
          at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
          at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
          at System.Data.SqlClient.SNI.SNITCPHandle.<ConnectAsync>d__22.MoveNext() in E:\work\gitclone\corefx\src\Syst
 em.Data.SqlClient\src\System\Data\SqlClient\SNI\SNITcpHandle.cs:line 191
          --- End of inner exception stack trace ---
          at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
          at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
          at System.Threading.Tasks.Task.Wait(TimeSpan timeout)
          at System.Data.SqlClient.SNI.SNITCPHandle..ctor(String serverName, Int32 port, Int64 timerExpire, Object cal
 lbackObject, Boolean parallel) in E:\work\gitclone\corefx\src\System.Data.SqlClient\src\System\Data\SqlClient\SNI\SNI
 TcpHandle.cs:line 145
       ---> (Inner Exception #0) System.ArgumentOutOfRangeException: Specified argument was out of the range of valid
 values.